pkg/skaffold/runner/runner.go

```diff
@@ -298,53 +302,62 @@ func (r *SkaffoldRunner) imageTags(ctx context.Context, out io.Writer, artifacts
 
 		select {
 		case <-ctx.Done():
-			return nil, context.Canceled
+			// TODO(nkubala): do we need to signal to the caller that this was cancelled?
+			return nil
 
 		case t := <-tagErrs[i]:
 			tag := t.tag
 			err := t.err
 			if err != nil {
-				return nil, errors.Wrapf(err, "generating tag for %s", imageName)
-			}
-
-			fmt.Fprintln(out, tag)
+				logrus.Errorf("error generating tag for image %s: build will be skipped", imageName)
+			} else {
+				fmt.Fprintln(out, tag)
 
-			imageTags[imageName] = tag
+				imageTags[imageName] = tag
+			}
 		}
 	}
 
 	color.Default.Fprintln(out, "Tags generated in", time.Since(start))
-	return imageTags, nil
+	return imageTags
 }
 
 // BuildAndTest builds artifacts and runs tests on built artifacts
-func (r *SkaffoldRunner) BuildAndTest(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) ([]build.Artifact, error) {
-	tags, err := r.imageTags(ctx, out, artifacts)
-	if err != nil {
-		return nil, errors.Wrap(err, "generating tag")
-	}
+func (r *SkaffoldRunner) BuildAndTest(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) ([]build.Result, error) {
+	tags := r.imageTags(ctx, out, artifacts)
 	r.hasBuilt = true
 
-	artifactsToBuild, res, err := r.cache.RetrieveCachedArtifacts(ctx, out, artifacts)
+	artifactsToBuild, res := r.cache.RetrieveCachedArtifacts(ctx, out, artifacts)
+
+	buildResultChannels, err := r.Build(ctx, out, tags, artifactsToBuild)
 	if err != nil {
-		return nil, errors.Wrap(err, "retrieving cached artifacts")
+		return nil, errors.Wrap(err, "failed to start build")
 	}
 
-	bRes, err := r.Build(ctx, out, tags, artifactsToBuild)
-	if err != nil {
-		return nil, errors.Wrap(err, "build failed")
+	bRes := build.CollectResultsFromChannels(buildResultChannels)
+
+	var errStr string
+	var buildErr error
+	for _, res := range bRes {
+		if res.Error != nil {
+			errStr += "\n" + res.Error.Error()
```

Comment:
I think it would be useful to add the image name here as well. Nothing guarantees that that's added in our errors unfortunately...wdyt?